specs/agents/tracing-api-otel.md

```diff
@@ -0,0 +1,251 @@
+
+if (netName && peerPort > 0) {
+    netName += ':';
+    netName += peerPort;
+}
+
+if (a['db.system']) {
+    type = 'db'
+    subtype = a['db.system'];
+    resource = netName || subtype;
+    if (a['db.name']) {
+        resource += '/'
+        resource += a['db.name'];
+    }
+
+} else if (a['messaging.system']) {
+    type = 'messaging';
+    subtype = a['messaging.system'];
+
+    if (!netName && a['messaging.url']) {
+        netName = parseNetName(a['messaging.url']);
+    }
+    resource = netName || subtype;
+    if (a['messaging.destination']) {
+        resource += '/';
+        resource += a['messaging.destination'];
+    }
+
+} else if (a['rpc.system']) {
+    type = 'external';
+    subtype = a['rpc.system'];
+    resource = netName || subtype;
+    if (a['rpc.service']) {
+        resource += '/';
+        resource += a['rpc.service'];
+    }
+
+} else if (a['http.url'] || a['http.scheme']) {
+    type = 'external';
+    subtype = 'http';
+
+    if (a['http.host'] && a['http.scheme']) {
+        resource = a['http.host'] + ':' + httpPortFromScheme(a['http.scheme']);
+    } else if (a['http.url']) {
+        resource = parseNetName(a['http.url']);
+    }
+}
+
+if (type === undefined) {
+    if (span.otel.span_kind == 'INTERNAL') {
+        type = 'app';
+        subtype = 'internal';
+    } else {
+        type = 'unknown';
+    }
+}
+span.type = type;
+span.subtype = subtype;
+span.destination.service.resource = resource;
+```
+
+### Active Spans and Context
+
+When possible, bridge implementation MUST ensure proper interoperability between Elastic transactions/spans and OTel spans when
+used from their respective APIs:
+- After activating an Elastic span via the agent's API, the [`Context`] returned via the [get current context API] should contain that Elastic span
+- When an OTel context is [attached] (aka activated), the [get current context API] should return the same [`Context`] instance.
+- Starting an OTel span in the scope of an active Elastic span should make the OTel span a child of the Elastic span.
+- Starting an Elastic span in the scope of an active OTel span should make the Elastic span a child of the OTel span.
+
+[`Context`]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/context.md
+[attached]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/context.md#attach-context
+[get current context API]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/context.md#get-current-context
+
+Both OTel and our agents have their own definition of what "active context" is, for example:
+- Java Agent: Elastic active context is implemented as a thread-local stack
+- Java OTel API: active context is implemented as a key-value map propagated through thread local
+
+In order to avoid potentially complex and tedious synchronization issues between OTel and our existing agent
```

Comment:
I ran into a little trickiness when implementing the "if there's not a parent span start a transaction" behavior when working on the Node.js Bridge related to the context objects.  I believe I have a path forward for the Node.js agent -- I'm mainly sharing this as background in case other folks run into it (and also forcing myself to say everything out loud in case there's some corner I'm not considering ;))

First -- there's no span builder object in any of the Node.js API or SDK code. Second, neither the `startSpan` or the `startActiveSpan` methods [of the Node.js tracer](https://github.com/open-telemetry/opentelemetry-js-api/blob/main/src/trace/tracer.ts) allow you to pass in the span that should be a parent.  Instead, this information is stored on the passed in open telemetry context object.  Specifically, the intended pattern here is that you can fetch the span that should be the parent (i.e. the span that's active) by using the context's `getSpanContext` method.

You can see an example of this in the [default Node.js SDK/tracer](https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-sdk-trace-base/src/Tracer.ts#L239-L244): 

The becomes a bit complicated because, currently, this span context is _initially set_ by whatever tracecontext propagator is currently being used

https://github.com/open-telemetry/opentelemetry-js/blob/04f9edd12fb2f42898bee7086691e47b3ab9b629/packages/opentelemetry-core/src/trace/W3CTraceContextPropagator.ts#L120

https://github.com/open-telemetry/opentelemetry-js/blob/04f9edd12fb2f42898bee7086691e47b3ab9b629/packages/opentelemetry-propagator-b3/src/B3SinglePropagator.ts#L86

https://github.com/open-telemetry/opentelemetry-js/blob/04f9edd12fb2f42898bee7086691e47b3ab9b629/packages/opentelemetry-propagator-b3/src/B3MultiPropagator.ts#L137

https://github.com/open-telemetry/opentelemetry-js/blob/04f9edd12fb2f42898bee7086691e47b3ab9b629/packages/opentelemetry-propagator-jaeger/src/JaegerPropagator.ts#L111

The trickiness came from the fact that we're not using these W3C propagators in the Node.js agent bridge -- trace context needs to be propagated when creating a transaction.  This means that in cases where we're starting a transaction and assigning it to the ElasticOtelSpan object, we also need to set that span context on the open telemetry context object.

Reply:
> This means that in cases where we're starting a transaction and assigning it to the ElasticOtelSpan object, we also need to set that span context on the open telemetry context object.

The idea for the bridges is that there's just a single context storage - the internal one of the agent. 

The case you mentioned is handled by this part of the spec:

> After activating an Elastic span via the agent's API, the [`Context`] returned via the [get current context API] should contain that Elastic span

What it means is that when [`ContextManager::active`](https://github.com/open-telemetry/opentelemetry-js-api/blob/b125324438fb2f24eb80c7c6673afc8cfc99575e/src/context/types.ts#L47) is called, the bridge implementation of the interface would create a `Context` on-the-fly which contains the Elastic span.


> The trickiness came from the fact that we're not using these W3C propagators in the Node.js agent bridge -- trace context needs to be propagated when creating a transaction. 

I don't think there's a need to implement a custom W3C propagators in the Node.js agent bridge. You can just rely on the provided implementation.

> The becomes a bit complicated because, currently, this span context is initially set by whatever tracecontext propagator is currently being used


The trace context propagators don't alter the context storage. It's a bit confusing but [`trace.setSpanContext`](https://github.com/open-telemetry/opentelemetry-js-api/blob/main/src/trace/context-utils.ts), which is [called by the propagators](https://github.com/open-telemetry/opentelemetry-js/blob/04f9edd12fb2f42898bee7086691e47b3ab9b629/packages/opentelemetry-core/src/trace/W3CTraceContextPropagator.ts#L120), only creates an immutable copy of the provided `Context` and sets the `span`. In the case of `W3CTraceContextPropagator`, the span is a "fake" `NonRecordingSpan`. The whole purpose of this is to populate a `Context` object that has the trace ids from the `traceparent` header.

Personally, I find that part of the OTel API a bit bloated and a source of unnecessary ceremony and allocation.

Happy to jump on a zoom if that was unclear.